up-l1/someip.adoc

```diff
@@ -0,0 +1,265 @@
+
+The following specification will elaborate on how uProtocol shall be mapped to *SOME/IP* protocol so that mechatronics applications and services can be connected to non-mechatronics and cloud based applications and services.
+
+=== Open Source Libraries
+
+As of the time of writing, there are two open source projects that are available for *SOME/IP*:
+
+1. *https://github.com/COVESA/vsomeip[COVESA-vsomeip]:* C++ solution based on the original specifications developed by BMW prior to contribution to AUTOSAR
+2. *https://projects.eclipse.org/projects/automotive.sommr[Eclipse-SommR]:*  Rust implementation based on the latest AUTOSAR *SOME/IP* specifications.
+
+NOTE: Given the latest version of *SOME/IP* specifications from AUTOSAR is not open source, we cannot share or implement uProtocol using Eclipse-SommR.
+
+=== AUTOSAR References
+All references below to:
+
+- `[PRS_SOMEIP_XXXXX]` refer to link:https://some-ip.com/standards.shtml[AUTOSAR *SOME/IP* specification].
+- `[PRS_SOMEIP_SD_XXXXX]` refer to link:https://some-ip.com/standards.shtml[SOME/IP Service Discovery Protocol Specification].
+
+A search engine usually finds the correct document based on the reference number.
+
+== Transport Layer Translation
+
+=== UAttributes
+
+The following table highlights the mapping of *uAttributes* to *SOME/IP* fields that apply to all event types. For more information about *uAttributes*, please refer to link:../basics/uattributes.adoc[uAttributes Specification].
+
+
+==== Uris
+
+link:../basics/uattributes.adoc[uAttributes Specification] explains that `source` attributes defines the address of whom sent the message, while `sink` defines the destination for the message in link:../basics/uri.adoc[`UUri`] format. `UUri` is translated to *SOME/IP* attributes per <<sink-source-uri-mapping>> below.
+
+.Sink & Source URI Mapping
+[sink-source-uri-mapping, cols="1,1,2"]
+|===
+| *UUri Portion* | *SOME/IP Field* | *Description*
+a| `authority_name` | IP address & port
+a| IP address & port (destination endpoint) of the mDevice.
+
+NOTE: `vsomeip` hides IP endpoint information and uses `Service ID`/`Instance ID` instead.
+
+a| `ue_id` a| `Instance ID` \| `Service ID`
+a| `Service ID` *MUST* be encoded in the lower 16 bits of `resource_id`.
+
+`Instance ID` *MAY* be encoded in the upper 16 bits of the 32 bit `ue_id`.
+If the upper 16 bits are `0x0000` (a reserved instance value), `Instance ID` is assumed to be `1`.
+
+NOTE: Explicitly setting `Instance ID` to `0x0001` *SHOULD* be avoided as it increases the message size.
+
+a| `ue_version_major` a| `Interface Version`
+a| `Interface Version` field of *SOME/IP Header*. `[PRS_SOMEIP_00053]`.
+
+Maps to `Major Version` for the `Service ID`. If set, it *SHOULD* be passed to underlying *SOME/IP* library to override it's default _Service Major Version_
+
+a| `resource_id` a| `Method ID`/`Event ID` | Identifier of the method/event placed in the
```

Comment:
As per my understanding,
uProtocol resource_id is mapped to method_id in someip header for RPC usecase and
uProtocol resource_id is mapped to event ID in someip header for Pub Sub usecase
Question: How are we going to map event group id in vsomeip ? If we se event_id same as event_groupID then we may have a problem. We faced this issue wirking with chimes team few weeks ago where they have usecase of different event_id and event_group ID for their publish-subscribe usecase.

Reply:
My opinion is still in favor of using `Event ID` for both `PUBLISH` and `SubscriptionRequest.topic.resource_id`.
This needs additional configuration for the transport (mapping Service/Instance/Event to a set of eventgroup IDs), but it allows a lot more flexibility in supporting ECUs.
The other option: forcing ECUs to have `EventID=EventGroupID` is a lot more restrictive for the HW and not OSS friendly. This can be assumed as "default" option - if no mapping is provided per event, assume `EventID=EventGroupID`